### PR TITLE
[kotlin][rfc] J2K: Add pre- and post-processing extension points

### PR DESCRIPTION
**Pull request by ermattt**
Based on testing this extension point with two dummy preprocessors (see [draft PR](https://github.com/ermattt/intellij-community/pull/45/files)), this works as expected (i.e. it runs the preprocessors on a background thread, and successfully applies their changes before conversion) with _one major caveat_: changes due to conversion and changes due to preprocessing require two undo actions (i.e. Ctl + z).

I've tried a couple things to fix this problem, but so far none has worked:
1. Running the custom preprocessors under a transparent global command ([code](https://github.com/ermattt/intellij-community/pull/47/files)).  Maybe I was supposed to enable merging of global commands somehow?
2. Using the same command group id to run the custom preprocessors and the committing new kotlin files + external code processing command ([code](https://github.com/ermattt/intellij-community/pull/46/files))

You can see the example preprocessors in action (plus the suboptimal undo behavior) in the attached video zip.
[Screen Recording 2024-06-24 at 11.08.57 AM.zip](https://github.com/user-attachments/files/15959012/Screen.Recording.2024-06-24.at.11.08.57.AM.zip)

Once the issues with the preprocessing extension point are resolved, I'll add a postprocessing extension point by more or less copying this one.

[KTIJ-29835](https://youtrack.jetbrains.com/issue/KTIJ-29835/J2K-Add-extension-point-for-custom-conversions-processings)

@abelkov @darthorimar @jocelynluizzi13

### CODE REVIEW

File: plugins/kotlin/j2k/k1.new.post-processing/src/org/jetbrains/kotlin/idea/j2k/post/processing/utils.kt [hunk 1]
```diff
@@ -6,7 +6,7 @@ import com.intellij.openapi.application.ApplicationManager
 import com.intellij.openapi.application.runWriteAction
 import com.intellij.openapi.command.CommandProcessor
 
-internal fun runUndoTransparentActionInEdt(inWriteAction: Boolean, action: () -> Unit) {
+fun runUndoTransparentActionInEdt(inWriteAction: Boolean, action: () -> Unit) {
```
Comment (darthorimar) [inline plugins/kotlin/j2k/k1.new.post-processing/src/org/jetbrains/kotlin/idea/j2k/post/processing/utils.kt:9]: This is an internal J2K function, let's not expose it :)

File: plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * `runUndoTransparentActionInEdt(inWriteAction = true) { ... }` so that they are executed on the EDT thread. As usual, read actions must
+ * be wrapped in `runReadAction { ... }`, and analysis must be done outside write actions.
+ */
+interface J2kPreprocessorExtension {
+
+    /**
+     * Override this method to analyze and edit Java files before conversion. This method is always executed on a background thread, so
+     * write actions must be wrapped in `runUndoTransparentActionInEdt(inWriteAction = true) { ... }`. As usual, read actions must be
+     * wrapped in `runReadAction { ... }`, and analysis must be done outside write actions.
+     */
```
Comment (darthorimar) [inline plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt:29]: Maybe in this case, it would be better to split `processFiles` into `processFiles` + `apply`? So the first one is executed from a BGT + read action while the second one is from the EDT + writeAction.

Another alternative would be to migrate this code to coroutines. There are `com.intellij.openapi.application.CoroutinesKt#writeAction` and `com.intellij.openapi.application.CoroutinesKt#readAction`. So the implementation of `processFiles` will look like

```
override suspend fun processFiles(project: Project, files: List<PsiJavaFile>) {`
  val result =   readAction { ... }
  writeAction { /* apply to the PSI tree here */ } // no explicit switching to EDT
}
```

File: plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/PreprocessorExtensionsRunner.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    fun runRegisteredPreprocessors(project: Project, javaFiles: List<PsiJavaFile>) {
+        val preprocessorExtensions = J2kPreprocessorExtension.EP_NAME.extensionList
+        if (preprocessorExtensions.isEmpty()) return
+
+        val preprocessorsCount = preprocessorExtensions.size
+        ProgressManager.progress(PHASE_NAME, "Found $preprocessorsCount preprocessors to run on Java files before conversion")
+
+        for ((i, preprocessor) in preprocessorExtensions.withIndex()) {
+            ProgressManager.checkCanceled()
+            ProgressManager.progress(PHASE_NAME, "Running preprocessor $i/$preprocessorsCount")
```
Comment (darthorimar) [inline plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/PreprocessorExtensionsRunner.kt:29]: The indexation here will be from `0`

**Review by darthorimar — CHANGES_REQUESTED** [review]

File: plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt (hunk 2 above)
Comment (ermattt) [reply plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt:33]: Done!

**Comment by ermattt** [comment]
Tests added and coroutines funkiness resolved!
I probably need to clean up some k1/k2 test duplicates, but other than that I think this is ready for review again. If it looks good to you, @darthorimar, I'll expand upon these changes to handle postprocessing as well.

**Comment by darthorimar** [comment]
@ermattt Awesome! Now tests behave more like production in `JavaToKotlinAction`.

The formatting changes in test data are definitely magic, but they seem to simulate the behavior in the real IntelliJ IDEA, so they are totally okay.

I'm ready to merge this PR as is, or do you want to have some modifications here?

>  If it looks good to you, @darthorimar, I'll expand upon these changes to handle postprocessing as well.
Definitely, thanks!

**Review by darthorimar — APPROVED** [review]

**Comment by ermattt** [comment]
I'll add my postprocessing changes to this PR today, test 'em, and then re-request review from you 🥳

File: plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt [hunk 4]
```diff
@@ -138,26 +149,39 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
...
-    open fun fileToKotlin(text: String, settings: ConverterSettings): String {
+    open fun fileToKotlin(text: String, settings: ConverterSettings, preprocessorExtensions: List<J2kPreprocessorExtension>): String {
         val file = createJavaFile(text)
         val j2kKind = if (isFirPlugin) K2 else K1_NEW
         val extension = J2kConverterExtension.extension(j2kKind)
         val converter = extension.createJavaToKotlinConverter(project, module, settings)
         val postProcessor = extension.createPostProcessor()
-        return converter.filesToKotlin(listOf(file), postProcessor).results.single()
+        var converterResult: FilesResult? = null
+        project.executeCommand("J2K") {
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:173]: We need this machinery for coroutines to work, right?

File: plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt [hunk 5]
```diff
@@ -138,26 +149,39 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
...
+                    converterResult =
+                        converter.filesToKotlin(listOf(file), postProcessor, preprocessorExtensions = preprocessorExtensions)
+                                                     }, "Testing J2K", /* canBeCanceled = */ true, project)
+        }
+        return checkNotNull(converterResult).results.single()
     }
 
     private fun methodToKotlin(text: String, settings: ConverterSettings): String {
-        val result = fileToKotlin("final class C {$text}", settings)
+        val result = fileToKotlin("final class C {$text}", settings, preprocessorExtensions = emptyList())
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:184]: Can be a default parameter value

File: plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt [hunk 6]
```diff
@@ -189,4 +213,21 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
 
     private fun createKotlinFile(text: String): KtFile =
         myFixture.configureByText("converterTestFile.kt", text) as KtFile
+
+    private val dummyPreprocessorEP = object : J2kPreprocessorExtension {
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:217]: Probably worth to extract to a top-level J2KTestPreprocessorExtension

File: plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ * The `org.jetbrains.kotlin.j2kPreprocessorExtension` extension point enables running custom preprocessing steps on Java files before they
+ * are converted to Kotlin. At runtime, all registered extensions are collected and executed sequentially. To implement your own
+ * preprocessor in a separate plugin, simply extend this interface and register the extension point in your plugin's xml file, e.g.
+ * ```
+ * <extensions defaultExtensionNs="org.jetbrains.kotlin">
+ *   <j2kPreprocessorExtension implementation="org.jetbrains.kotlin.j2k.FooPreprocessorExtension"/>
+ * </extensions>
+ * ```
+ *
+ * All preprocessors are run on a background thread, so write actions must be wrapped in
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt:19]: Probably worth to mention that they are run in coroutines and therefore must use coroutine-compatible API

File: plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ * `writeAction { ... }` so that they are executed on the EDT thread. As usual, read actions must
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt:20]: FQNs for writeAction/readAction here would be nice to avoid confusion

File: plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt [hunk 9]
```diff
@@ -189,4 +213,21 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
...
+        override suspend fun processFiles(
+            project: Project,
+            files: List<PsiJavaFile>,
+        ) {
+            for (file in files) {
+                val method = readAction {
+                    file.classes.firstOrNull()?.findDescendantOfType<PsiMethod> { it.name != "main" && !it.isConstructor && !it.name.startsWith("get") && !it.name.startsWith("set") }
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:224]: long line

File: plugins/kotlin/j2k/shared/tests/testData/newJ2k/annotations/annotationUsages.k1.kt [hunk 10]
```diff
@@ -18,7 +18,8 @@ import javaApi.E
 @Anon5(1)
 @Anon6("x", "y")
 @Anon7(
-    String::class, StringBuilder::class
+    String::class,
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/tests/testData/newJ2k/annotations/annotationUsages.k1.kt:21]: Formatting changes are because we now run test J2K inside of a command? Does this indeed reflect production J2K results?

**Review by abelkov — CHANGES_REQUESTED** [review]

File: plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * Preprocessors are only applied to file-level conversions; copy-paste conversions ignore registered preprocessors.
+ *
+ * Internally, preprocessors can use either K1 or K2 utilities.
+ */
+interface J2kPreprocessorExtension {
+
+    /**
+     * Override this method to analyze and edit Java files before conversion. This method is always executed on a background thread, so
+     * write actions must be wrapped in `runUndoTransparentActionInEdt(inWriteAction = true) { ... }`. As usual, read actions must be
```
[the remark this answers is not in the document]
Comment (ermattt) [reply plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/j2k/J2kPreprocessorExtension.kt]: My bad, that was an outdated kdoc!

File: plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt (hunk 4 above)
Comment (ermattt) [reply plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:179]: Unfortunately yes

File: plugins/kotlin/j2k/shared/tests/testData/newJ2k/annotations/annotationUsages.k1.kt (hunk 10 above)
Comment (ermattt) [reply plugins/kotlin/j2k/shared/tests/testData/newJ2k/annotations/annotationUsages.k1.kt:21]: So it seems! I have no idea why, but it lines up with the results i get if I try converting a similar expression in a test instance of the IDE

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/actions/JavaToKotlinAction.kt [hunk 12]
```diff
@@ -82,7 +79,12 @@ class JavaToKotlinAction : AnAction() {
                 val progressIndicator = ProgressManager.getInstance().progressIndicator!!
 
                 val conversionTime = measureTimeMillis {
-                    converterResult = converter.filesToKotlin(javaFiles, postProcessor, progressIndicator)
+                    converterResult = converter.filesToKotlin(
+                        javaFiles,
+                        postProcessor,
+                        progressIndicator,
+                        preprocessorExtensions = J2kPreprocessorExtension.EP_NAME.extensionList
```
[the remark this answers is not in the document]
Comment (ermattt) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/actions/JavaToKotlinAction.kt]: It's still too long for one line when I include postprocessors too :(

**Comment by ermattt** [comment]
@abelkov & @darthorimar, addressed comments and tested the example pre- and post-processor extensions together (see screen recording)
[preAndPostTogether.mov.zip](https://github.com/user-attachments/files/16189086/preAndPostTogether.mov.zip)

File: plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt [hunk 13]
```diff
@@ -190,3 +222,45 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
...
+            } ?: continue
+
+            writeAction {
+                setName(checkNotNull(method.nameIdentifier), "prebar")
+            }
+        }
+    }
+}
+
+private val dummyPostprocesorEP = object : J2kPostprocessorExtension {
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:245]: Let's move them to testUtils.kt and make them internal named objects?

File: plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/preAndPostprocessorExtensionsRunner.kt [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+            } catch (e: ProcessCanceledException) {
+                throw e
+            } catch (t: Throwable) {
+                Logger.getInstance(J2kExtensionsRunner::class.java).error(t)
+                commitAllDocuments(files)
+            }
+        }
+    }
+
+    private fun commitAllDocuments(javaFiles: List<T>) {
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/preAndPostprocessorExtensionsRunner.kt:58]: `files`

File: plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt [hunk 15]
```diff
@@ -138,22 +149,43 @@ abstract class AbstractJavaToKotlinConverterSingleFileTest : AbstractJavaToKotli
...
+    ): String {
         val file = createJavaFile(text)
         val j2kKind = if (isFirPlugin) K2 else K1_NEW
         val extension = J2kConverterExtension.extension(j2kKind)
         val converter = extension.createJavaToKotlinConverter(project, module, settings)
         val postProcessor = extension.createPostProcessor()
-        return converter.filesToKotlin(listOf(file), postProcessor).results.single()
+        var converterResult: FilesResult? = null
+        project.executeCommand("J2K") {
+            ProgressManager.getInstance()
```
Comment (abelkov) [inline plugins/kotlin/j2k/shared/tests/test/org/jetbrains/kotlin/j2k/AbstractJavaToKotlinConverterSingleFileTest.kt:177]: A bit nicer formatting:

```
val process = {
    converterResult = converter.filesToKotlin(
        listOf(file),
        postProcessor,
        EmptyProgressIndicator(),
        preprocessorExtensions,
        postprocessorExtensions
    )
}
project.executeCommand("J2K") {
    ProgressManager.getInstance().runProcessWithProgressSynchronously(process, "Testing J2K", /* canBeCanceled = */ true, project)
}
```

**Review by abelkov — CHANGES_REQUESTED** [review]

**Comment by ermattt** [comment]
@abelkov just addressed the last round of comments

**Review by abelkov — APPROVED** [review]

**Review by darthorimar — APPROVED** [review]
